Add valid_after and valid_before parameters to SSH certificate signing

alex-sherwin opened the pull request:
adds optional `valid_after` and `valid_before` parameters to the sign
SSH key API

this change can be summarized as:

1. existing behavior remains the same
2. valid_after and valid_before cannot be used if ttl is used
3. valid_after and valid_before must be used together
4. the period between valid_after and valid_before must be <= max_ttl
5. valid_before cannot be before valid_after

The intent here is to have a 1:1 feature parity with the ability to specify the validity period of the SSH certificate to be signed

Currently, a SSH certificate can only be signed from `[now - 30s]` to `[now + ttl]`

alex-sherwin commented on builtin/logical/ssh/path_sign.go:
```diff
@@ -355,17 +460,24 @@ func (b *backend) calculateExtensions(data *framework.FieldData, role *sshRole)
+//  2. true if the returned TTL was specified in the data, false otherwise
+//  3. calculated maxTTL
+//  4. any error
+func (b *backend) calculateTTL(data *framework.FieldData, role *sshRole) (time.Duration, time.Duration, bool, error) {
```
Modified the return signature here to return some pertinent information so the certificate sign function can make some new decisions:

1. whether `ttl` was specified or derived from roles/defaults (to be used to check if `ttl` was specified at the same time as `valid_after` and `valid_before`)
1. what the calculated `max_ttl` was (to be used to check if the period of time between `valid_after` and `valid_before` exceeds `max_ttl`

alex-sherwin commented on builtin/logical/ssh/path_sign.go:
```diff
@@ -28,7 +28,8 @@ type creationBundle struct {
 	CertificateType uint32
-	TTL             time.Duration
+	ValidBefore     uint64
+	ValidAfter      uint64
```
replaced `TTL` here with `ValidAfter` and `ValidBefore`, the logic to calculate them was pulled up into the sign function.

this is done because the sign function holds the information necessary to decide if `ValidAfter` and `ValidBefore` should be calculated based on the `ttl` (specified or derived), or, if they are to be given explicitly by the new `valid_after` and `valid_before` parameters

alex-sherwin commented on builtin/logical/ssh/path_sign.go:
```diff
@@ -172,12 +213,28 @@ func (b *backend) pathSignCertificate(ctx context.Context, req *logical.Request,
+		// calculate certificate ValidAfter/ValidBefore using TTL
+		now := time.Now()
+		certValidAfter = uint64(now.Add(-30 * time.Second).In(time.UTC).Unix())
+		certValidBefore = uint64(now.Add(ttl).In(time.UTC).Unix())
```
This is the original logic to derive `ValidAfter` and `ValidBefore` from the `ttl`, it was pulled up into the sign function since there are now two ways these values can be set

alex-sherwin commented on builtin/logical/ssh/path_sign.go:
```diff
@@ -468,15 +580,13 @@ func (b *creationBundle) sign() (retCert *ssh.Certificate, retErr error) {
 		KeyId:           b.KeyID,
 		ValidPrincipals: b.ValidPrincipals,
-		ValidAfter:      uint64(now.Add(-30 * time.Second).In(time.UTC).Unix()),
-		ValidBefore:     uint64(now.Add(b.TTL).In(time.UTC).Unix()),
```
This logic was pulled up into the sign function

alex-sherwin commented on builtin/logical/ssh/backend_test.go:
```diff
@@ -105,6 +105,13 @@ SeKWrUkryx46LVf6NMhkyYmRqCEjBwfOozzezi5WbiJy6nn54GQt
 `
 )
 
+type CertValidityValidationMethod int
```
This type is used to branch the validation logic in the `validateSSHCertificate` function.

alex-sherwin commented on builtin/logical/ssh/backend_test.go:
```diff
@@ -105,6 +105,13 @@ SeKWrUkryx46LVf6NMhkyYmRqCEjBwfOozzezi5WbiJy6nn54GQt
+type CertValidityValidationMethod int
+
+const (
+	ValidateTTL            CertValidityValidationMethod = 0
```
This `ValidateTTL` enum is used to apply the existing validation logic for signed SSH certificates whose `ValidAfter` and `ValidBefore` were derived from `ttl`

alex-sherwin commented on builtin/logical/ssh/backend_test.go:
```diff
@@ -105,6 +105,13 @@ SeKWrUkryx46LVf6NMhkyYmRqCEjBwfOozzezi5WbiJy6nn54GQt
+
+const (
+	ValidateTTL            CertValidityValidationMethod = 0
+	ValidateAfterAndBefore CertValidityValidationMethod = 1
```
This `ValidateAfterAndBefore` enum is used to apply new logic for validation of the `ValidAfter` and `ValidBefore` values when they are specified by the new `valid_after` and `valid_before` parameters

alex-sherwin commented on builtin/logical/ssh/backend_test.go:
```diff
@@ -916,17 +1326,31 @@ func validateSSHCertificate(cert *ssh.Certificate, keyID string, certType int, v
 
-	actualTTL := time.Unix(int64(cert.ValidBefore), 0).Add(-30 * time.Second).Sub(time.Unix(int64(cert.ValidAfter), 0))
-	if actualTTL != ttl {
-		return fmt.Errorf("incorrect ttl: expected: %v, actualL %v", ttl, actualTTL)
```
The typo in `actualL` was fixed to `actual`

alex-sherwin commented on builtin/logical/ssh/path_sign.go:
```diff
@@ -172,12 +213,28 @@ func (b *backend) pathSignCertificate(ctx context.Context, req *logical.Request,
+		certValidAfter = validAfter
+		certValidBefore = validBefore
+	} else {
+		// calculate certificate ValidAfter/ValidBefore using TTL (specified or derived)
```
The default conditional logic here is to do the previous behavior (use the TTL [specified or derived] to calculate the ValidAfter/ValidBefore values)

alex-sherwin commented on builtin/logical/ssh/path_sign.go:
```diff
@@ -172,12 +213,28 @@ func (b *backend) pathSignCertificate(ctx context.Context, req *logical.Request,
+
+	// branch the logic for calculating the certificate ValidAfter & ValidBefore values
+	if validAfterWasSpecified && validBeforeWasSpecified {
+		// the ValidAfter/ValidBefore were specified as epoch's directly in the specified data
```
Only apply the new logic to directly set ValidAfter/ValidBefore if they were both specified, we know that other logical checks already took place to ensure that `ttl` was not explicitly specified

alex-sherwin commented:
Pinging @jefferai , @vishalnayak , @briankassouf , @kalafut 

Sorry I'm not sure if you have a protocol for community contributed PR's, but I'm just wondering if there's any appetite to adopt this change or if I will need to maintain my own builds with this feature long-term

Thanks,

tyrannosaurus-becks commented:
Hi @alex-sherwin , thanks for working on this! I'm not sure whether it's the direction we'll want to go, but we certainly appreciate your time.

I was curious about a couple things. One, what is the use case for adding these fields? Two, can that use case be achieved by using the current code?

I lean towards avoiding the change if at all possible because 1) it adds complexity, and 2) it reduces this plugin's similarity to other ones, which makes it harder to learn and use. However, I'm not _entrenched_ in leaning away from this change, given that I lack information on its purpose.

alex-sherwin commented:
Hi @tyrannosaurus-becks, thanks for looking into this PR

The OpenSSH certificate spec (https://github.com/openssh/openssh-portable/blob/master/PROTOCOL.certkeys) is defined with `valid after` and `valid before` epochs, so a signed certificate can be created for any time span.

The current Vault implementation unfortunately only allows for a TTL concept, which does not align 1:1 with the OpenSSH certificate spec.  The current logic hard-codes the `valid after` to `[now] - [30 seconds]` and `valid before` to `[now] + [TTL]`.

This PR leaves the current TTL logic in place, but augments with a second code path for directly specifying `valid after` & `valid before` which can only be utilized when the TTL parameter is not explicitly provided in the sign certificate API request.

Conceptually, allowing direct specification of `valid after` and `valid before`, as this PR does, now  allows for any possible use-case for OpenSSH certificates.

Whereas the currently hard-coded TTL mechanism is acting as an opinionated & inflexible middle-man on top of what the underlying spec is actually capable of.

The reason I made this change & PR was to utilize Vault as the primary crypto utility for an in-house PAM (privileged access management) system. 

Vault nicely covered 100% of the crypto needs, between PKI, on-demand symmetric encryption and OpenSSH signed certificates.

The only problem I had was the system needed the ability to provision  time-boxed credentials ahead of time.

For example, a DevOps user or team may be planning to do Production maintenance work for a 4 hour time span on the Weekend.  We want to generate an access request during regular working hours, and have it go through a managerial approval/sign-off workflow ahead of time.  We want to be able to then generate signed SSH certificates for the time-boxed 4 hour window in the future.

Without this change, the DevOps user/team will need to wait until the 4 hour window begins, and then have Vault generate the certificate.  But this route could be potentially problematic, because it's now a timing exercise for our PAM system to only execute the Vault API certificate signing API inside that window, and each time we do so we would need to dynamically calculate what the new TTL should be (for example, if the DevOps user waited 15 minutes into the 4 hour window, when we execute the Vault API we would need to set the TTL to 3 hours 45 minutes).

By being able to simply specify explicit `valid after` and `valid before` certificate epoch's, the alignment between our PAM and Vault's behavior is greatly simplified and eliminates the possibility of any kind of timing mis-alignment.  It also alleviates the requirement that the Vault interactions can only occur within that time span.

As for the code changes in the PR, I strived to make the absolute minimum necessary changes to support this, with good test coverage, which I think could be summarized in the following points:

* pulled `ValidAfter` and `ValidBefore` up to the `creationBundle` struct
* pulled the existing logic for calculating the TTL-based `ValidAfter`/`ValidBefore` from the `sign` function up to the `pathSignCertificate` so that it would be co-located with the other logic for validation and value calculation
   * also, importantly, allows for this PR to have two code paths for deciding `ValidAfter`/`ValidBefore` and then passing them into `sign`
* updated `calculateTTL` to return more values, namely, `ttlWasSpecified`, which was necessary so that it could be detected if `ttl` was explicitly specified in the API request (as opposed to derived from defaults)
* added functions to calculate `ValidAfter`/`ValidBefore` when those new API parameters are utilized
* modified the `pathSignCertificate` business logic to
   1. by default, everything behaves as it did before this PR change
   1. don't allow usage of new `valid_after` and `valid_before` API params if `ttl` was explicitly present
   1. apply usage constraints on `valid_after` and `valid_before` (must be specified as a pair, `valid_after` must be before `valid_before` and the time span between them must not exceed Vaults derived `max_ttl`)
* added appropriate test coverage

I would really love to see this PR make it into the code base, I think it nicely opens up any potential OpenSSH signed certificate use-case (future proofing and positioning Vault as an invaluable tool in this space).

There are not a lot of tools aside from OpenSSH itself which currently support dealing with OpenSSH signed certificates (not even libssh/libssh2).  But, I think knowledge of and adoption of signed certificates is going to start increasing and I think this PR can make Vault the only good tool out there with a programmatic HTTP API capable of 1:1 feature parity with the OpenSSH certificate spec.

I'm more then happy to refactor something that you perceive as a problem to adoption, or add more test coverage if there are scenarios or variations I hadn't considered yet

Sorry for the long post, but thank you for your time!

tyrannosaurus-becks commented:
@alex-sherwin - thank you for both the technical and usability reasoning for this code! I think that's very valid and now do favor adding this PR. I will look at it more thoroughly and will either approve it or provide feedback, with an aim to including it in the code base. Should be able to do that within the week, if not sooner.

tyrannosaurus-becks commented:
I notice there are two test failures in this directory; however, they're identical master too so I'm not going to worry related to this PR. They haven't been noticed because they're tests that aren't run by CI. Just documenting them as being unaffected by this code. I'll open a separate issue for them.
```
--- FAIL: TestSSHBackend_DynamicKeyCreate (51.64s)
--- FAIL: TestSSHBackend_CredsForZeroAddressRoles_dynamic (0.21s)
```

tyrannosaurus-becks commented on builtin/logical/ssh/path_sign.go:
```diff
@@ -48,6 +49,14 @@ func pathSign(b *backend) *framework.Path {
 			},
+			"valid_after": &framework.FieldSchema{
+				Type:        framework.TypeInt,
+				Description: `The "valid after" UNIX epoch in seconds, must be used with valid_before.  Cannot be used with ttl.`,
```
```suggestion
				Description: `The "valid after" UNIX epoch in seconds, must be used with valid_before. Cannot be used with ttl.`,
```

tyrannosaurus-becks commented on builtin/logical/ssh/path_sign.go:
```diff
@@ -48,6 +49,14 @@ func pathSign(b *backend) *framework.Path {
+			},
+			"valid_before": &framework.FieldSchema{
+				Type:        framework.TypeInt,
+				Description: `The "valid before" UNIX epoch in seconds, must be used with valid_after.  Cannot be used with ttl.`,
```
```suggestion
				Description: `The "valid before" UNIX epoch in seconds, must be used with valid_after. Cannot be used with ttl.`,
```

tyrannosaurus-becks commented on builtin/logical/ssh/path_sign.go:
```diff
@@ -48,6 +49,14 @@ func pathSign(b *backend) *framework.Path {
 				Description: `The desired role with configuration for this request.`,
 			},
+			"valid_after": &framework.FieldSchema{
+				Type:        framework.TypeInt,
```
I see why you've picked a type int to receive the epoch seconds. It matches the spec you've linked.

We're very big on usability for the Vault API, and we also tend to heavily standardize our fields. I wasn't able to find any precedents for API fields in Vault that represent a specific point in time. That means these two fields might inadvertently create a precedent for how the Vault API consumes date fields! :-) 

I had always imagined that we would tend to get and give date fields that would be in the [ISO-8601](https://en.wikipedia.org/wiki/ISO_8601) format due to both their _human readability_ and their interoperability. I'm pretty sure `2019-07-08T16:39:51Z` would be a forward-compatible date-time format should we ever want to create a framework field for that. 

Would you be willing to switch this over to being a string that only accepts dates and times in that format?

alex-sherwin replied:
Funnily enough, I initially started implementing this from the top down (Vault API first) as ISO-8601 strings, and when I finally got down to the struct passed to the OpenSSH signing lib saw that it was unix epoch's, I just rolled it all back and used int's to represent an epoch.

I'd be find with switching it to a text based ISO 8601 string, if you could point me to an example of the way you would expect to parse ISO 8601 strings with Go (Go is not my primary development language, so I don't know what the de-facto standard is for standards based time parsing)

tyrannosaurus-becks replied:
Sure! Thank you! Go does it in an unusual way. [Here](https://golang.org/pkg/time/#Parse) is the "official" tome, and in a nutshell, it would be:
```
whateverVar, err := time.Parse("2006-01-02T15:04:05Z", userValueString)
```

You might want to double-check me, but I believe that's it.

tyrannosaurus-becks commented on builtin/logical/ssh/path_sign.go:
```diff
@@ -204,6 +261,54 @@ func (b *backend) pathSignCertificate(ctx context.Context, req *logical.Request,
+//  3. any error
+func (b *backend) calculateValidAfter(data *framework.FieldData) (uint64, bool, error) {
+
+	validAfterRaw, valueWasSet, err := data.GetOkErr("valid_after")
```
Not a big deal, but FWIW, it would be possible to just use `GetOK` here and eliminate the type check lines on 273 thru 275. The outer framework will handle and error if it's given input that doesn't parse as the given schema type.

tyrannosaurus-becks commented on builtin/logical/ssh/path_sign.go:
```diff
@@ -204,6 +261,54 @@ func (b *backend) pathSignCertificate(ctx context.Context, req *logical.Request,
+//  3. any error
+func (b *backend) calculateValidBefore(data *framework.FieldData) (uint64, bool, error) {
+
+	validBeforeRaw, valueWasSet, err := data.GetOkErr("valid_before")
```
Ditto here.

ncabatoff commented:
Hi @alex-sherwin,

Did you still want to pursue this PR?  It looks like the only thing holding it back right now is the use of epoch time in the API (well, that and some merge conflicts).

alex-sherwin commented:
Hi @ncabatoff,

Yes I would, sorry this fell by the wayside with work for quite some time

I've got a vacation coming up shortly, I'd like to take a pass at refactoring to use epoch time
